channeldb/migration23/migration.go

```diff
@@ -0,0 +1,128 @@
+	// info about the attempt that was done for the HTLC in question.
+	oldAttemptInfoKey = []byte("htlc-attempt-info")
+
+	// oldSettleInfoKey is a key used in a HTLC's sub-bucket to store the
+	// settle info, if any.
+	oldSettleInfoKey = []byte("htlc-settle-info")
+
+	// oldFailInfoKey is a key used in a HTLC's sub-bucket to store
+	// failure information, if any.
+	oldFailInfoKey = []byte("htlc-fail-info")
+
+	// htlcAttemptInfoKey is the key used as the prefix of an HTLC attempt
+	// to store the info about the attempt that was done for the HTLC in
+	// question. The HTLC attempt ID is concatenated at the end.
+	htlcAttemptInfoKey = []byte("ai")
+
+	// htlcSettleInfoKey is the key used as the prefix of an HTLC attempt
+	// settle info, if any. The HTLC attempt ID is concatenated at the end.
+	htlcSettleInfoKey = []byte("si")
+
+	// htlcFailInfoKey is the key used as the prefix of an HTLC attempt
+	// failure information, if any.The  HTLC attempt ID is concatenated at
+	// the end.
+	htlcFailInfoKey = []byte("fi")
+)
+
+// htlcBucketKey creates a composite key from prefix and id where the result is
+// simply the two concatenated. This is the exact copy from payments.go.
+func htlcBucketKey(prefix, id []byte) []byte {
+	key := make([]byte, len(prefix)+len(id))
+	copy(key, prefix)
+	copy(key[len(prefix):], id)
+	return key
+}
+
+// MigrateHtlcAttempts will gather all htlc-attempt-info's, htlcs-settle-info's
+// and htlcs-fail-info's from the attempt ID buckes and re-store them using the
+// flattened keys to each payment's payment-htlcs-bucket.
+func MigrateHtlcAttempts(tx kvdb.RwTx) error {
+	paymentsBucket := tx.ReadWriteBucket(paymentsRootBucket)
+	if paymentsBucket == nil {
+		return nil
+	}
+
+	return paymentsBucket.ForEach(func(k, v []byte) error {
+		// Get the bucket which contains the payment, fail if the key
+		// does not have a bucket.
+		bucket := paymentsBucket.NestedReadWriteBucket(k)
+		if bucket == nil {
+			return fmt.Errorf("non bucket element in '%v' bucket",
+				paymentsRootBucket)
+		}
+
+		htlcsBucket := bucket.NestedReadWriteBucket(paymentHtlcsBucket)
+		if htlcsBucket == nil {
+			// Nothing to migrate for this payment.
+			return nil
+		}
+
+		return migrateHtlcsBucket(htlcsBucket)
+	})
+}
+
+// migrateHtlcsBucket is a helper to gather, transform and re-store htlc attempt
+// key/values.
+func migrateHtlcsBucket(htlcsBucket kvdb.RwBucket) error {
+	kvs := make(map[string][]byte)
+
+	if err := htlcsBucket.ForEach(func(aid, v []byte) error {
+		attemptBucket := htlcsBucket.NestedReadWriteBucket(aid)
+
+		if attemptBucket == nil {
+			return fmt.Errorf("non bucket element in '%v' bucket",
+				paymentHtlcsBucket)
+		}
+
+		// Collect attempt/settle/fail infos.
+		attemptInfo := attemptBucket.Get(oldAttemptInfoKey)
+		if attemptInfo != nil {
+			key := string(htlcBucketKey(htlcAttemptInfoKey, aid))
+			kvs[key] = attemptInfo
+		}
+
+		settleInfo := attemptBucket.Get(oldSettleInfoKey)
+		if settleInfo != nil {
```

Comment:
Do we need to care about `nil` vs. empty slice here, depending on the backend used? Maybe use `len() > 0` instead? They should never be empty if they exist, right?